projects/certs/X509_certificate_to_der.cc

```diff
@@ -0,0 +1,180 @@
+}
+
+void CertToDER::Encode(const VersionNumber& version) {
+  // |version| is Context-specific with tag number 0 (RFC 5280, 4.1 & 4.1.2.1).
+  // Takes on values 0, 1 and 2, so only require length of 1 to
+  // encode it (RFC 5280, 4.1 & 4.1.2.1).
+  std::vector<uint8_t> der_ver_num = {0x80, 0x01, static_cast<uint8_t>(version)};
+  der_.insert(der_.end(), der_ver_num.begin(), der_ver_num.end());
+}
+
+void CertToDER::Encode(const TBSCertificateValue& tbs_certificate) {
+  // Save the current size in |tag_len_pos| to place sequence tag and length
+  // after the value is encoded.
+  size_t tag_len_pos = der_.size();
+
+  Encode(tbs_certificate.version());
+  Encode(tbs_certificate.serial_number());
+  Encode(tbs_certificate.signature_algorithm());
+  Encode(tbs_certificate.issuer().name());
+  Encode(tbs_certificate.validity());
+  Encode(tbs_certificate.subject().name());
+  Encode(tbs_certificate.subject_public_key_info());
+
+  // Preserve to later backtrack and explicitly set tag.
+  size_t pos_of_tag;
+  // RFC 5280, 4.1: |issuer_unique_id| and |subject_unique_id|
+  // are only set for v2 and v3 and |extensions| only set for v3.
+  // However, set |issuer_unique_id|, |subject_unique_id|, and |extensions|
+  // independently of the version number for interesting inputs.
+  if (tbs_certificate.has_issuer_unique_id()) {
+    pos_of_tag = der_.size();
+    Encode(tbs_certificate.issuer_unique_id().unique_identifier());
+    // |issuer_unqiue_id| is Context-specific with tag number 1 (RFC 5280, 4.1
+    // & 4.1.2.8).
+    der_[pos_of_tag] = 0x81;
+  }
+  if (tbs_certificate.has_subject_unique_id()) {
+    pos_of_tag = der_.size();
+    Encode(tbs_certificate.subject_unique_id().unique_identifier());
+    // |subject_unqiue_id| is Context-specific with tag number 2 (RFC 5280, 4.1
+    // & 4.1.2.8).
+    der_[pos_of_tag] = 0x82;
+  }
+  if (tbs_certificate.has_extensions()) {
+    pos_of_tag = der_.size();
+    Encode(tbs_certificate.extensions());
+    // |extensions| is Context-specific with tag number 3 (RFC 5280, 4.1
+    // & 4.1.2.8).
+    der_[pos_of_tag] = 0x83;
+  }
+
+  // The fields of |tbs_certificate| are wrapped around a sequence (RFC
+  // 5280, 4.1 & 4.1.2.5).
+  // Sequence is constructed and has tag number 16 (X.208, Table 1).
+  // The current size of |der_| subtracted by |tag_len_pos|
+  // equates to the size of the value of |tbs_certificate|.
+  EncodeTagAndLength(0x30, der_.size() - tag_len_pos, tag_len_pos, der_);
+}
+
+template <typename T>
+void CertToDER::Encode(const T& t) {
+  if (t.has_pdu()) {
+    Encode(t.pdu());
+    return;
+  }
+  Encode(t.value());
+}
+
+std::vector<uint8_t> CertToDER::X509CertificateToDER(
+    const X509Certificate& X509_certificate) {
+  // Reset the previous state.
+  der_.clear();
+
+  // Save the current size in |tag_len_pos| to place sequence tag and length
+  // after the value is encoded.
+  size_t tag_len_pos = der_.size();
```

Comment:
```suggestion
  size_t tag_len_pos = 0;
```

You just cleared it on 161 ;)